add in memory linking

banditopazzo opened the pull request:
I ended my first iteration for memory linking and api changes.

To recap my goals:
1. in memory linking
2. and since I am touching the external apis, add the possibility of reusing the linker instance avoiding a full initialization each time (in my use case I have to basically link in a for loop)

Major changes:
- better/safer wrappers for llvm objects (custom `Drop` in every wrapper, lifetimes, mutability)
- api change to have 2 main methods: link_to_file, link_to_buffer
- linker accepts both files and in-memory data at the same time
- moved some linker methods to standalone functions

Open points or missing:
- from what I understand the target_machine could be created early and stored in the linker instance, in theory it's based on the cli options
- names of the new structures
- `dump_module` should be moved to link methods

I am not surprised if I'm forgetting something. 

In any case this should be enough at least to have an idea of changes I am proposing

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aya-rs/bpf-linker/295)
<!-- Reviewable:end -->

banditopazzo commented:
another question is what to do with `libs` path list: to remove or to handle

vadorovsky reviewed (commented):
Do you think you could split out the wrapper parts as a separate change? If so, we could merge it fast, and then review just the in-memory part of the code (which I still need to play with and review properly).

alessandrod commented:
> Hi @alessandrod , have you had a chance to take a look at this draft? I'd appreciate your thoughts on whether the changes work for you and if you like the direction it’s heading in.

gonna take a look today, sorry for the delay!

alessandrod commented on src/linker.rs:
```diff
@@ -226,345 +268,84 @@ pub struct LinkerOptions {
+    /// Link and generate the output code to file.
+    pub fn link_to_file(
+        &self,
+        inputs: Vec<LinkerInput>,
```
nit: this should probably be `&[LinkerInput]`

alessandrod commented on src/linker.rs:
```diff
@@ -226,345 +268,84 @@ pub struct LinkerOptions {
-        self.create_target_machine()?;
+    fn link<'ctx>(
+        &'ctx self,
+        inputs: Vec<LinkerInput>,
```
same here, slice?

alessandrod commented on src/linker.rs:
```diff
@@ -633,3 +414,333 @@ fn detect_input_type(data: &[u8]) -> Option<InputType> {
+
+fn link_modules<'ctx>(
+    context: &'ctx LLVMContextWrapped,
+    inputs: Vec<LinkerInput>,
```
slice?

tamird replied:
yeah, slice - or better yet, `IntoIter<Item=Path>`
<!-- Reviewable comment -ObdgSzO6Tslx0sAw_5q:b-gjw63z -->
<!-- Sent from Reviewable.io -->

alessandrod commented on src/linker.rs:
```diff
@@ -226,345 +268,84 @@ pub struct LinkerOptions {
+
+    pub fn link_to_buffer(
+        &self,
+        inputs: Vec<LinkerInput>,
```
slice?

alessandrod commented on src/linker.rs:
```diff
@@ -226,345 +268,84 @@ pub struct LinkerOptions {
+        inputs: Vec<LinkerInput>,
+        output: &Path,
+        output_type: OutputType,
+        export_symbols: HashSet<Cow<'static, str>>,
```
take by ref?

alessandrod commented on src/linker.rs:
```diff
@@ -633,3 +414,333 @@ fn detect_input_type(data: &[u8]) -> Option<InputType> {
+    inner: MemoryBufferWrapped,
+}
+
+impl LinkerOutput {
```
should probably also implement `AsRef<[u8]>`

alessandrod commented on src/llvm/module.rs:
```diff
@@ -0,0 +1,74 @@
+}
+
+impl<'ctx> LLVMModuleWrapped<'ctx> {
+    pub unsafe fn get_target(&self) -> *const i8 {
```
should be as `*const c_char`

alessandrod commented on examples/basic.rs:
```diff
@@ -0,0 +1,54 @@
+use bpf_linker::{Cpu, Linker, LinkerInput, LinkerOptions, OptLevel, OutputType};
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let path = Path::new("/path/to/object-or-bitcode");
```
wouldn't this example be better as Examples in the rustdocs for the respective methods?

banditopazzo commented:
I rebased on the main and squashed to remove useless commits.

The I addressed most of the review comments.

@vadorovsky: I tried to create a new PR without the in memory linking but it ended just like this, the in memory linking is marginal. I think it doesn't make the review process faster, but if really want as separate PR, I will do it.

@alessandrod: taking a slice instead of the `Vec<LinkerInput>` as input wouldn't work in my opinion. The `LinkerInput` objects have a mutable state inside because of the cursor they use for the `Read`, so it needs to be a `&mut [LinkerInput]`. I don't see any advantage doing this, considering `LinkerInput` is just a type that we use to accept a mixed vector of both files and in memory data

alessandrod commented on src/bin/bpf-linker.rs:
```diff
@@ -118,10 +118,6 @@ struct CommandLine {
     #[clap(long)]
     allow_bpf_trap: bool,
 
-    /// Add a directory to the library search path
```
why this change? bpf-linker needs to stay compatible with rustc, we can't change
the CLI

banditopazzo replied:
ok got it, it is because of rustc, I will revert it

I removed it because it was unused

alessandrod commented on src/linker.rs:
```diff
@@ -229,354 +266,169 @@ pub struct LinkerOptions {
+    ///
+    /// // Link to a file
+    /// linker.link_to_file(
+    ///     vec![
```
this is why Vec is bad: you're forcing people to allocate here, while

```
[
	LinkerInput::try_from(path),
]
```

should work too

something like ```impl IntoIterator<Item = impl BorrowMut<LinkerInput<'a>>>```
should work

alessandrod commented on src/linker.rs:
```diff
@@ -141,6 +137,58 @@ pub enum OptLevel {
+    file: File,
+}
+
+pub struct BytesInput<'a> {
```
nit: I'd call this BufferInput, FileInput contains bytes too

alessandrod commented on src/linker.rs:
```diff
@@ -141,6 +137,58 @@ pub enum OptLevel {
+    }
+}
+
+impl<'a> From<(&'a str, &'a [u8])> for LinkerInput<'a> {
```
This doesn't look great. Now that there are two ways to create LinkerInput, I'd
make it new_from_file and new_from_buffer

alessandrod commented on src/linker.rs:
```diff
@@ -645,3 +497,347 @@ fn detect_input_type(data: &[u8]) -> Option<InputType> {
+    }
+}
+
+impl AsRef<[u8]> for LinkerOutput {
```
Deref?

vadorovsky commented on src/llvm/types/target_machine.rs:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    pub unsafe fn codegen_to_mem(
```
Unnecessary unsafe?

vadorovsky commented on src/linker.rs:
```diff
@@ -141,6 +137,58 @@ pub enum OptLevel {
 
+pub struct FileInput<'a> {
+    path: &'a Path,
+    file: File,
```
I think it would be better to keep only the `path` here and open the file when we are actually processing the input.

Currently, because of storing the `File` instance in `FileInput`, the code in `bpf-linker.rs` looks like:

```rust
    let inputs: Vec<LinkerInput> = inputs
        .iter()
        .map(|p| LinkerInput::try_from(p.as_path()))
        .collect::<Result<_, _>>()?;
```

It has to do an allocation just to handle the `Result`, where error can happen if a file cannot be opened.

If we:

* keep just a `Path` in `FileInput`
* handle opening the file somewhere else
* apply @alessandrod's comment about taking `impl IntoIterator<Item = impl BorrowMut<LinkerInput<'a>>>`

that code could become:

```rust
    let inputs: Vec<LinkerInput> = inputs
        .iter()
        .map(|p| LinkerInput::try_from(p.as_path()));
```

alessandrod commented on src/linker.rs:
```diff
@@ -229,354 +288,183 @@ pub struct LinkerOptions {
+    ///
+    /// // Link into an in-memory buffer.
+    /// let out_buf = linker.link_to_buffer(
+    ///     vec![
```
nit: vec not necessary anymore

alessandrod commented on src/linker.rs:
```diff
@@ -229,354 +288,183 @@ pub struct LinkerOptions {
+        &self,
+        inputs: impl IntoIterator<Item = LinkerInput<'i>>,
+        output_type: OutputType,
+        export_symbols: &HashSet<Cow<'static, str>>,
```
any reason export_symbols and dump_module are passed here and not as part of
LinkerOptions? 

Making inputs and output_type arguments here make sense if we add support for
emitting multiple formats (eg you might want to emit both .ir and .asm for debugging)

alessandrod replied:
> the reason is because now the `Linker` object is reusable, so the symbols to export may change depending on the needs. instead regarding `dump_module`, you don't want to overwrite the pre and post optimization files if you call `link_` functions multiple times.
> 
> for example my use case is something like this:
> 
> ```rust
> let linker = Linker::new(LinkerOptions { 
>   // ...
> })?;
> 
> for item in stuff {
>   // ...
>   let export_symbols = todo!(); // calculate export symbols
>   let dump_module_path = todo!(); // optional dump path
>   // ...
>   let result = linker.link_to_buffer(
>         inputs,
>         output_type,
>         &export_symbols,
>         dump_module_path.as_deref(),
>   )?;
>   // ...
> }
> ```
> 
> the symbols I have to export are different for every iteration and I want a different path for each `link_*` call in case I have to look at the ir.

Yes, but you can also do this by doing 
```
linker.options.dump_module_path = todo!();
let res = linker.link_to_buffer(inputs, output_type, exported_symbols);
```

When designing an API, you have to think what's the most common case, and what's the cognitive overhead of leaking more advanced fringe cases to the common API. You have convinced me that passing `export_symbols` to link is correct, but `dump_module_path` is definitely a fringe case, and shouldn't pollute the API. 

Also in your example you've made it `&HashSet::new()` which is incorrect - if you don't export any symbols DCE will remove all the code, please fix that. 

> I agree that maybe it should be improved, but right now this PR is too big, I think it's better to split in other tasks.
> 
> I have few other stuff I want to add but after this

This is the central point of this PR, and it's a public API, so we have to get it as right as possible and we should aim to break API as little as possible.

alessandrod commented on src/llvm/types/context.rs:
```diff
@@ -0,0 +1,78 @@
+        self.context
+    }
+
+    pub fn create_module<'ctx>(&'ctx self, name: &str) -> Option<LLVMModule<'ctx>> {
```
nit: API wise seems a bit weird that we have LLVMContext::new() but instead of
having LLVMModule::new() we have LLVMContext::create_module

vadorovsky replied:
I also prefer `create_module` method. This API feels intuitive to me and makes it clear that modules are created only inside contexts, never just on their own out of nowhere.

LLVM C++ API implements [an `LLVMContext` constructor that takes `LLVMContext &C` as an argument](https://llvm.org/doxygen/classllvm_1_1Module.html#a378f93ece2ac999e500f07056cfe6528), but I actually find it confusing and inconsistent, because then all other entities inside `LLVMContext` and `Module` are created by methods like:

* [`LLVMContext::getOrInsertBundleTag`](https://llvm.org/doxygen/classllvm_1_1LLVMContext.html#a624ccf92d0c47b9a2da6534870b007b0)
* [`LLVMContext::getOrInsertSyncScopeID`](https://llvm.org/doxygen/classllvm_1_1LLVMContext.html#adc484415b469ae5d438c3fd803ca3a4f)
* [`Module::getOrInsertFunction`](https://llvm.org/doxygen/classllvm_1_1Module.html#abb107e4edbf05eae92936cba6801c2d9)
* [`Module::getOrInsertGlobal`](https://llvm.org/doxygen/classllvm_1_1Module.html#ae2fcaead977b30daccedc4e397d5e1f0)
* [`Module::getOrInsertNamedMetadata`](https://llvm.org/doxygen/classllvm_1_1Module.html#ab33e5587b25f9c80f6ea5000124d1a2b)

etc.

alessandrod commented on src/llvm/types/module.rs:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    pub fn write_ir_to_memory(&self) -> MemoryBuffer {
+        // NOTE: This function implementation wraps the result into a MemoryBuffer,
```
I'd change this comment to:

```
// Format the module to a string, then copy into a MemoryBuffer. We do the extra copy to keep the
// internal API simpler, as all the other codegen methods output a MemoryBuffer.
```

alessandrod commented on src/llvm/types/target_machine.rs:
```diff
@@ -0,0 +1,110 @@
+        self.target_machine
+    }
+
+    pub fn codegen_to_path(
```
nit: call this emit_to_file like the underlying LLVM API?

alessandrod commented on src/llvm/types/target_machine.rs:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    pub fn codegen_to_mem(
```
nit: name emit_to_memory_buffer like the underlying API?

alessandrod reviewed (commented):
Sorry for the delay! I just finished my crazy month of traveling, so I'll be more responsive from now.

It's looking pretty good, just a couple of API comments

vadorovsky commented:
No comments besides what @alessandrod said. The wrappers look good now, thanks for addressing my previous comments. :slightly_smiling_face: 

There is a huge chance that the CI failures will disappear after you rebase (beta switched to newer LLVM, 1.86.0 looks like a flake).

vadorovsky commented:
@banditopazzo I pushed a commit with a fix. The problem was that the C strings were not null terminated and apparently LLVM 19 expects the null character. It's a good practice to add it for all FFI calls anyways. Sorry for suggesting a faulty path conversion in my previous comments!

alessandrod reviewed (changes requested):
See https://github.com/aya-rs/bpf-linker/pull/295#discussion_r2384601821

That's the only thing left then it's ready to go

vadorovsky commented on src/linker.rs:
```diff
@@ -645,3 +533,354 @@ fn detect_input_type(data: &[u8]) -> Option<InputType> {
+    info!("LLVM command line: {:?}", args);
+
+    unsafe {
+        llvm::init(&args, "BPF linker");
+    }
+
+    let context = LLVMContext::new();
+    context.set_diagnostic_handler(&mut diagnostic_handler);
```
You could use `Pin` and `Box` for the diagnostic handler, like @tamird did here:

https://github.com/aya-rs/bpf-linker/pull/307/files#diff-2ae8d7b12653fc1f1413b12c532f72291f5caeea0afa0257d31091b03015eafaR235

Given that your PR is older, required a lot of work and is pretty close to being accepted by everyone, I think it's better to get in yours first. Pinning is probably the only critical issue that linters found in @tamird's PR.

vadorovsky replied:
Uh, ok, that PR actually got merged, sorry :/

tamird commented:
I landed quite a large change in #307 which might be a pain to rebase, sorry about that. On the bright side it removes many unwraps.

tamird commented:
The merge conflict here is unfortunately very large due to all the code movement included in this PR. IMO it would be better to separate functional changes from code movement both to ameliorate this problem and to ease code review.

tamird commented on src/llvm/mod.rs:
```diff
@@ -65,18 +64,8 @@ pub(crate) fn init(args: &[Cow<'_, CStr>], overview: &CStr) {
     unsafe { LLVMParseCommandLineOptions(c_ptrs.len() as i32, c_ptrs.as_ptr(), overview.as_ptr()) };
 }
 
-pub(crate) fn create_module(name: &CStr, context: LLVMContextRef) -> Option<LLVMModuleRef> {
```
here is an example of an API I had changed (to accept CStr) to avoid an unwrap, which is lost in this rebase.

banditopazzo replied:
yeah, it was very difficult to rebase, I had to constantly look at your changes. I have to double check everything, this was my first round. 

And I would like to also find a safer solution for the diagnostic handler. Pin locks the memory address but doesn't ensure it remains alive

vadorovsky commented on src/llvm/types/module.rs:
```diff
@@ -0,0 +1,85 @@
+
+    pub unsafe fn write_ir_to_file(&self, output: &CStr) -> Result<(), String> {
+        let (ret, message) =
+            Message::with(|message| LLVMPrintModuleToFile(self.module, output.as_ptr(), message));
```
Hmm... in the current shape, where all the `write_*` and `emit_*` methods are being called only in [`codegen_to_file`](https://github.com/aya-rs/bpf-linker/blob/eda9d591613984846975b66daaf4af6f6d915cab/src/linker.rs#L733-L754), we could indeed to the conversion there, once for all output types:

```rust
fn codegen_to_file(
    module: &LLVMModule<'_>,
    target_machine: &LLVMTargetMachine,
    output: &Path,
    output_type: OutputType,
) -> Result<(), LinkerError> {
    info!("writing {:?} to {:?}", output_type, output);
    let output = CString::new(output
        .as_os_str()
        .as_encoded_bytes()
        .as_ptr()
        .cast());
    match output_type {
        OutputType::Bitcode => module
            .write_bitcode_to_path(output)
            .map_err(LinkerError::WriteBitcodeError),
        OutputType::LlvmAssembly => module
            .write_ir_to_path(output)
            .map_err(LinkerError::WriteIRError),
        OutputType::Assembly => target_machine
            .emit_to_file(module, output, LLVMCodeGenFileType::LLVMAssemblyFile)
            .map_err(LinkerError::EmitCodeError),
        OutputType::Object => target_machine
            .emit_to_file(module, output, LLVMCodeGenFileType::LLVMObjectFile)
            .map_err(LinkerError::EmitCodeError),
    }
}
```

I think at the time I was making that comment, the calls of these functions were spread in different places and the `CString` construction was done there, overlaping with the actual logic too much.

But honestly, I'm also fine with leaving as it is.
<!-- Reviewable comment -Oc0xj3h6EPCdLqmjPDg:b-5pjc0l -->
<!-- Sent from Reviewable.io -->

tamird replied:
@banditopazzo would you mind doing this please? this is all private API, so i think it would be better to have one fewer unwrap.
<!-- Reviewable comment -Oc1sf573DjhttyEJTS7:b4mddo3 -->
<!-- Sent from Reviewable.io -->

banditopazzo replied:
Done.
<!-- Reviewable comment -Oc5sUoo1uOrhoBQiLa7:b-896fix -->
<!-- Sent from Reviewable.io -->

tamird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/aya-rs/bpf-linker/295)* status: all files reviewed, 6 unresolved discussions (waiting on @alessandrod, @banditopazzo, and @vadorovsky)



<!-- Sent from Reviewable.io -->

banditopazzo reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/aya-rs/bpf-linker/295)* status: all files reviewed, 6 unresolved discussions (waiting on @alessandrod, @tamird, and @vadorovsky)

___
*[`Cargo.toml` line 37 at r66](https://reviewable.io/reviews/aya-rs/bpf-linker/295#-Oc1p-1z0TvaaZpa_qxu:-Oc2-Ygf2cJJIuYa59oW:b-896fix) ([raw file](https://github.com/aya-rs/bpf-linker/blob/8610dcd283e3546b894bf10aa4b33b7b455b4d55/Cargo.toml#L37)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

let's not bring in a dependency for this
</blockquote></details>

Done.
___
*[`src/linker.rs` line 690 at r70](https://reviewable.io/reviews/aya-rs/bpf-linker/295#-Oc1p_Ay8IvmYO5Q__nL:-Oc2-sHV9Ggo2tOZux5s:byakpug) ([raw file](https://github.com/aya-rs/bpf-linker/blob/62b4466caf6d4d0f9ed4554a591cc94d8f3bbc54/src/linker.rs#L690)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this...does not need to be owned.
</blockquote></details>

I think it's not possible with the previous function signature. The only way I found to avoid the allocation is to change the signature taking `IntoIterator<Item = &'a str>,`
___
*[`src/bin/bpf-linker.rs` line 274 at r70](https://reviewable.io/reviews/aya-rs/bpf-linker/295#-Oc1pnzP7zso4Ni4v1Zt:-Oc2-cg0EZ87rw_ZAjAj:b-896fix) ([raw file](https://github.com/aya-rs/bpf-linker/blob/62b4466caf6d4d0f9ed4554a591cc94d8f3bbc54/src/bin/bpf-linker.rs#L274)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think you just resolved this TODO.
</blockquote></details>

Done.
___
*[`src/llvm/types/module.rs` line 42 at r64](https://reviewable.io/reviews/aya-rs/bpf-linker/295#-Oc1HQSj49cmhuyLsIni:-Oc2-Tt47X4Zl7bwH_ps:b-896fix) ([raw file](https://github.com/aya-rs/bpf-linker/blob/f9a5b4d234e39f86ea72694ee1ec6acf7a8f190e/src/llvm/types/module.rs#L42)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

https://github.com/lambda-fairy/rust-errno?tab=readme-ov-file#comparison-with-stdioerror

can we just use https://doc.rust-lang.org/std/io/struct.Error.html#method.last_os_error and return `std::io::Error` instead of `String`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

vadorovsky commented:
I've restarted the failing job, it it fails everytime with:

```
cargo:warning=error: failed to write /tmp/rustczv1z8A/lib.rmeta: No space left on device (os error 28)
cargo:warning=
cargo:warning=
error: could not compile `rustix` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
cargo:warning=error: failed to build archive at `/home/runner/work/bpf-linker/bpf-linker/aya/target/release/deps/libaya_obj-dee67e733d80af3e.rlib`: No space left on device (os error 28)
cargo:warning=
cargo:warning=
error: could not compile `aya-obj` (lib) due to 1 previous error
rustc-LLVM ERROR: IO failure on output stream: No space left on device
```

And the latest build on `main` also complains about the lack of space:

https://github.com/aya-rs/bpf-linker/actions/runs/18672411715/job/53235931949

I'll try to fix in a separate PR, but I think it's safe to merge this one, given that all the other jobs also run the integration tests.

vadorovsky commented on src/linker.rs:
```diff
@@ -232,367 +271,558 @@ pub struct LinkerOptions {
+    output_type: OutputType,
+) -> Result<(), LinkerError> {
+    info!("writing {:?} to {:?}", output_type, output);
+    let output = CString::new(output.as_os_str().as_encoded_bytes()).unwrap();
```
Meh, ignore, let's do it separately

alessandrod commented on src/linker.rs:
```diff
@@ -232,367 +271,558 @@ pub struct LinkerOptions {
+    ///
+    /// let export_symbols = ["my_sym_1", "my_sym_2"];
+    ///
+    /// linker.link_to_file(
```
lmao I'm so dumb I swear apparently I accidentally put the emoji there while reviewing!? SORRY
